app/background.cc

```diff
@@ -24,45 +23,45 @@ struct background_t {
+    auto proc = last;
     while (pbkd->run) {
-        // read frame under lock
-        pbkd->mux.lock();
-        if (pbkd->cap.read(pbkd->raw)) {
-            pbkd->frm += 1;
-            // grab timing point and calculate frame period
-            struct timespec now;
-            clock_gettime(CLOCK_REALTIME, &now);
-            double nsec = (double)(now.tv_sec-last.tv_sec)*1e9 + (double)(now.tv_nsec-last.tv_nsec);
-            last = now;
+        // read new frame - we use a temporary buffer for two reasons:
+        // - we can read unlocked, thus we are not impacted by blocking backends (eg: V4L2)
+        // - we preserve the last frame for callers, such that if a video ends, they always have a frame to use
+        cv::Mat grab;
+        if (pbkd->cap.read(grab)) {
+            // got a frame, lock and copy for callers
+            {
+                std::unique_lock<std::mutex> hold(pbkd->mux);
+                grab.copyTo(pbkd->raw);
+                pbkd->frm += 1;
+            }
+            // grab timing point
+            auto now = std::chrono::steady_clock::now();
             // display thumbnail frame with overlay info if debug enabled
             if (pbkd->dbg) {
                 char msg[40];
-                snprintf(msg, sizeof(msg), "FPS:%0.1f FRM:%d", 1e9/nsec, pbkd->frm);
+                long nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(now-last).count();
+                snprintf(msg, sizeof(msg), "FPS:%0.1f FRM:%d", 1e9/(double)nsec, pbkd->frm);
                 cv::Mat frame;
-                cv::resize(pbkd->raw, frame, cv::Size(240,160));
+                cv::resize(grab, frame, cv::Size(240,160));
                 cv::putText(frame, msg, cv::Point(5,frame.rows-5), cv::FONT_HERSHEY_PLAIN, 1.0, cv::Scalar(255,255,255));
                 cv::imshow("Background", frame);
             }
-            pbkd->mux.unlock();
+            last = now;
             // wait for next frame, some sources are real-time, others are not, this ensures all play in real-time.
             if (pbkd->frm>1) {
-                long adj = (now.tv_sec-proc.tv_sec)*1000000000l + (now.tv_nsec-proc.tv_nsec);
-                now.tv_sec = 0;
-                now.tv_nsec= (long)(1.0/pbkd->fps*1e9);
-                // adjust for processing time or skip if less than that
-                if (now.tv_nsec > adj) {
-                    now.tv_nsec -= adj;
-                    nanosleep(&now, nullptr);
-                } else {
-                    if (pbkd->dbg) fprintf(stderr, "background: wait=%ld adj=%ld\n", now.tv_nsec, adj);
+                auto adj = std::chrono::duration_cast<std::chrono::nanoseconds>(now-proc);
+                auto del = std::chrono::nanoseconds((long)(1e9/pbkd->fps));
+                // adjust for processing time and sleep or skip if no time left
+                if (del > adj) {
+                    del -= adj;
+                    std::this_thread::sleep_for(del);
                 }
```

Comment:
Prone to overall jitter AFAIC. Especially as the `std::this_thread::sleep_for` does not guarantee exact timings.

Reply:
Any thoughts on what to use from `std::` to improve the timing? I don't think there are any underlying implementations that provide guarantees on resolution, hence nothing in the std library can either?

Reply:
I've introduced some min/max measures, and switched to absolute target time via `std::this_thread::sleep_util()` calculated from frame rate. This doesn't show visible jitter (to me anyway!) with a smoothly moving video (eg: https://file-examples-com.github.io/uploads/2020/03/file_example_WEBM_1920_3_7MB.webm). would still welcome thoughts :smile: